### ISSUE TITLE
Segfault on 1.1.2 whenever running PHP

### ISSUE THREAD

**Issue by Zenexer** [mongodb/mongo-php-driver#209]
On Ubuntu 15.10 x64:

```
% sudo php5-fpm -t || echo "Exit status: $?"
[27-Jan-2016 01:35:47] NOTICE: configuration file /etc/php5/fpm/php-fpm.conf test is successful

Exit status: 139
```

Here's an excerpt from catchsegv:

```
Backtrace:
php5-fpm[0x6dee88]
/usr/lib/x86_64-linux-gnu/sasl2/libdigestmd5.so(+0x391b)[0x7f29ffd7491b]
/usr/lib/x86_64-linux-gnu/libsasl2.so.2(+0x6cc9)[0x7f2a07ba5cc9]
/usr/lib/x86_64-linux-gnu/libsasl2.so.2(sasl_client_done+0x2a)[0x7f2a07ba5e0a]
/usr/lib/php5/20131226/mongodb.so(+0x6122e)[0x7f2a003eb22e]
/lib/x86_64-linux-gnu/libpthread.so.0(+0xe36b)[0x7f2a0d3c036b]
/lib64/ld-linux-x86-64.so.2(+0x10bb7)[0x7f2a0fb96bb7]
/lib/x86_64-linux-gnu/libc.so.6(+0x39d32)[0x7f2a0d609d32]
/lib/x86_64-linux-gnu/libc.so.6(+0x39d85)[0x7f2a0d609d85]
php5-fpm(fpm_init+0x182)[0x7b9912]
php5-fpm(main+0x776)[0x465366]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0)[0x7f2a0d5f0a40]
php5-fpm(_start+0x29)[0x466539]
```

If I remove package libsasl2-dev before compiling, it works fine.

**Comment by jmikola** [comment]
Sorry for losing track of this issue. Was this using the `mongodb` extension installed through PECL or a separate package repository? Also, do you recall what exact versions of PHP and libsasl you were using at the time?

Fwiw, your stack trace looks a bit like an older segfault I encountered months ago (see: [this comment](https://jira.mongodb.org/browse/PHPLIB-117?focusedCommentId=1020596&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-1020596)), although I had more frames visible. If this is reproducible, is there any chance you can get a more detailed backtrace (e.g. `bt full` via `gdb`) from PHP with debug symbols installed and/or compiling the extension with:

```
phpize && ./configure --enable-developer-flags && make all && make install
```

**Comment by EugenMayer** [comment]
As Derik pointed out, its related to https://github.com/mongodb/mongo-php-driver/issues/258

Uninstalling libsasl2-dev and recompiling the extension using PECL: pecl install mongodb does indeed fix this.

The reason why most people have libsasl2-dev installed is, since memcached-dev depends on it, which is needed to install pecl install memached.

After reinstalling mongodb like described above, valgrind is clean:

valgrind --leak-check=full /usr/bin/php -S localhost:8085
==19935== Memcheck, a memory error detector
==19935== Copyright (C) 2002-2011, and GNU GPL'd, by Julian Seward et al.
==19935== Using Valgrind-3.7.0 and LibVEX; rerun with -h for copyright info
==19935== Command: /usr/bin/php -S localhost:8085
==19935==
PHP 5.5.33-1~dotdeb+7.1 Development Server started at Thu Mar 10 13:04:48 2016
Listening on http://localhost:8085
Document root is /root
Press Ctrl-C to quit.
^C==19935==
==19935== HEAP SUMMARY:
==19935==     in use at exit: 109,428 bytes in 353 blocks
==19935==   total heap usage: 37,385 allocs, 37,032 frees, 5,233,203 bytes allocated
==19935==
==19935== LEAK SUMMARY:
==19935==    definitely lost: 0 bytes in 0 blocks
==19935==    indirectly lost: 0 bytes in 0 blocks
==19935==      possibly lost: 0 bytes in 0 blocks
==19935==    still reachable: 109,428 bytes in 353 blocks
==19935==         suppressed: 0 bytes in 0 blocks
==19935== Reachable blocks (those to which a pointer was found) are not shown.
==19935== To see them, rerun with: --leak-check=full --show-reachable=yes
==19935==
==19935== For counts of detected and suppressed errors, rerun with: -v
==19935== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 30 from 8)

**Comment by Zenexer** [comment]
@jmikola Straight from PECL on Ubuntu 15.10.  I didn't do anything fancy.  At the time I was using libsasl2 2.0.25, I believe, although that may be a more recent version.

**Comment by Zenexer** [comment]
The problem appears to be a *little* better with mongodb 1.1.4 from PECL and libsasl2 2.1.26.  PHP-FPM isn't instantly crashing on every request now.  Seeing almost identical output from catchsegv:

```
php5-fpm[0x6dee88]
/usr/lib/x86_64-linux-gnu/sasl2/libdigestmd5.so(+0x391b)[0x7faa0704491b]
/usr/lib/x86_64-linux-gnu/libsasl2.so.2(+0x6cc9)[0x7faa0ee76cc9]
/usr/lib/x86_64-linux-gnu/libsasl2.so.2(sasl_client_done+0x2a)[0x7faa0ee76e0a]
/usr/lib/php5/20131226/mongodb.so(+0x6159e)[0x7faa076bb59e]
/lib/x86_64-linux-gnu/libpthread.so.0(+0xe36b)[0x7faa1469136b]
/lib64/ld-linux-x86-64.so.2(+0x10c07)[0x7faa16e68c07]
/lib/x86_64-linux-gnu/libc.so.6(+0x39d32)[0x7faa148dad32]
/lib/x86_64-linux-gnu/libc.so.6(+0x39d85)[0x7faa148dad85]
php5-fpm(fpm_init+0x182)[0x7b9912]
php5-fpm(main+0x776)[0x465366]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0)[0x7faa148c1a40]
php5-fpm(_start+0x29)[0x466539]
```

`make test` failed; I'll assume either my environment is wrong or the tests aren't being maintained:

```
Number of tests :  229               228
Tests skipped   :    1 (  0.4%) --------
Tests warned    :    0 (  0.0%) (  0.0%)
Tests failed    :  212 ( 92.6%) ( 93.0%)
Expected fail   :    2 (  0.9%) (  0.9%)
Tests passed    :   14 (  6.1%) (  6.1%)
```

The list of failures is far too long to include here; I can drop it in gist if you'd like.

Here's an excessively verbose trace from gdb:

```
#0  _zend_mm_free_int (heap=0xf2ef00, p=0x1025c00) at /build/php5-qbYJ0w/php5-5.6.11+dfsg/Zend/zend_alloc.c:2104
        mm_block = 0x1025bf0
        next_block = 0x2055980
        size = 16973200
#1  0x00007fffe886005e in php_phongo_free (mem=<optimized out>) at /home/paul/pecl-mongodb/mongodb-1.1.4/php_phongo.c:2342
No locals.
#2  0x00007fffe81cb91b in ?? () from /usr/lib/x86_64-linux-gnu/sasl2/libdigestmd5.so
No symbol table info available.
#3  0x00007fffefffccc9 in ?? () from /usr/lib/x86_64-linux-gnu/libsasl2.so.2
No symbol table info available.
#4  0x00007fffefffce0a in sasl_client_done () from /usr/lib/x86_64-linux-gnu/libsasl2.so.2
No symbol table info available.
#5  0x00007fffe884185e in _mongoc_do_cleanup () at /home/paul/pecl-mongodb/mongodb-1.1.4/src/libmongoc/src/mongoc/mongoc-init.c:130
No locals.
#6  0x00007ffff581736b in __pthread_once_slow (once_control=0x7fffe8a97d30 <once>, init_routine=0x7fffe8841850 <_mongoc_do_cleanup>) at pthread_once.c:114
        _buffer = {__routine = 0x7ffff58173c0 <clear_once_control>, __arg = 0x7fffe8a97d30 <once>, __canceltype = 80, __prev = 0x0}
        val = 0
        newval = <optimized out>
#7  0x00007ffff7de9c07 in _dl_fini () at dl-fini.c:252
        array = 0x7fffe8a8a908
        i = 1
        nmaps = 107
        nloaded = <optimized out>
        i = 81
        l = 0x100a330
        ns = 0
        maps = 0x7fffffffb9b0
        maps_size = 856
        do_audit = <optimized out>
        __PRETTY_FUNCTION__ = "_dl_fini"
#8  0x00007ffff5a60d32 in __run_exit_handlers (status=status@entry=0, listp=0x7ffff5deb698 <__exit_funcs>, run_list_atexit=run_list_atexit@entry=true) at exit.c:82
        atfct = <optimized out>
        onfct = <optimized out>
        cxafct = <optimized out>
        f = <optimized out>
#9  0x00007ffff5a60d85 in __GI_exit (status=status@entry=0) at exit.c:104
No locals.
#10 0x00000000007b9912 in fpm_init (argc=argc@entry=2, argv=argv@entry=0x7fffffffe678, config=<optimized out>, prefix=prefix@entry=0x0, pid=pid@entry=0x0, test_conf=test_conf@entry=1, 
    run_as_root=0, force_daemon=-1, force_stderr=0) at /build/php5-qbYJ0w/php5-5.6.11+dfsg/sapi/fpm/fpm/fpm.c:72
        __func__ = "fpm_init"
#11 0x0000000000465366 in main (argc=2, argv=0x7fffffffe678) at /build/php5-qbYJ0w/php5-5.6.11+dfsg/sapi/fpm/fpm/fpm_main.c:1941
        exit_status = 0
        c = -1
        use_extended_info = 0
        file_handle = {type = ZEND_HANDLE_FILENAME, filename = 0x0, opened_path = 0x0, handle = {fd = 0, fp = 0x0, stream = {handle = 0x0, isatty = 0, mmap = {len = 0, pos = 0, map = 0x0, 
                buf = 0x0, old_handle = 0x0, old_closer = 0x0}, reader = 0x0, fsizer = 0x0, closer = 0x0}}, free_filename = 0 '\000'}
        orig_optind = 1
        orig_optarg = 0x0
        ini_entries_len = 0
        max_requests = 500
        requests = 0
        fcgi_fd = 0
        request = {listen_socket = 0, fd = 0, id = 0, keep = 0, closed = 0, in_len = 0, in_pad = 0, out_hdr = 0x0, out_pos = 0x0, out_buf = '\000' <repeats 3400 times>..., 
          reserved = "x\346\377\377\377\177\000\000p\377\336\367\377\177\000", env = 0x1}
        fpm_config = 0x0
        fpm_prefix = 0x0
        fpm_pid = 0x0
        test_conf = 1
        force_daemon = -1
        force_stderr = 0
        php_information = 0
        php_allow_to_run_as_root = 0
        __func__ = "main"
```

Relevant code:

``` c
static MONGOC_ONCE_FUN( _mongoc_do_cleanup)
{
#ifdef MONGOC_ENABLE_SSL
   _mongoc_ssl_cleanup();
#endif

#ifdef MONGOC_ENABLE_SASL
#ifdef MONGOC_HAVE_SASL_CLIENT_DONE
   sasl_client_done ();  /* Line 130; this is where it's segfaulting */
#else
   /* fall back to deprecated function */
   sasl_done ();
#endif
#endif

#ifdef _WIN32
   WSACleanup ();
#endif

   _mongoc_counters_cleanup ();

   MONGOC_ONCE_RETURN;
}
```

**Comment by Zenexer** [comment]
Here's the code from frame 2:

``` c
static void gssapi_common_mech_free(void *global_context __attribute__((unused)),
                    const sasl_utils_t *utils)
{
#ifdef GSS_USE_MUTEXES
    if (gss_mutex) {
      utils->mutex_free(gss_mutex);  /* SEGV */
      gss_mutex=NULL;
    }
#endif
}
```

And then we follow the breadcrumbs:

```
(gdb) p utils->mutex_free
$5 = (sasl_mutex_free_t *) 0x7fffe88417f0 <mongoc_sasl_mutex_free>
```

``` c
/* mongo-init.c */
static void
mongoc_sasl_mutex_free (void *mutex)
{
   mongoc_mutex_destroy ((mongoc_mutex_t *) mutex);
   bson_free (mutex);  /* SEGV */
}
```

```
/* bson-memory.c */
void
bson_free (void *mem) /* IN */
{
   gMemVtable.free (mem);  /* SEGV */
}
```

Which brings us back to `php_phongo_free`.

mongoc_mutex_destroy is a macro for [pthread_mutex_destroy](http://linux.die.net/man/3/pthread_mutex_destroy).

**Comment by Zenexer** [comment]
I could be completely misunderstanding this, but why is bson_free being called?  If a mutex managed via the mongoc_* methods is really just a pthread mutex without any sort of wrapper, it doesn't make sense to free it through libbson.

**Comment by Zenexer** [comment]
zend_alloc.c:

```
static void _zend_mm_free_int(zend_mm_heap *heap, void *p ZEND_FILE_LINE_DC ZEND_FILE_LINE_ORIG_DC)
{
    /*
    (gdb) p *heap
    $6 = {use_zend_alloc = 1, _malloc = 0x0, _free = 0x0, _realloc = 0x0, free_bitmap = 0, large_free_bitmap = 131072, block_size = 262144, compact_size = 2097152, segments_list = 0x7ffff7f1e010, 
      storage = 0xf2eee0, real_size = 262144, real_peak = 262144, limit = 25165824, size = 9120, peak = 9120, reserve_size = 8192, reserve = 0x7ffff7f1e030, overflow = 0, internal = 0, cached = 856, 
      cache = {0x7ffff7f202d8, 0x7ffff7f20378, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x7ffff7f20068, 0x0 <repeats 53 times>, 0x7ffff7f200c8, 0x0}, free_buckets = {0xf2f188, 0xf2f188, 0xf2f198, 0xf2f198, 
        0xf2f1a8, 0xf2f1a8, 0xf2f1b8, 0xf2f1b8, 0xf2f1c8, 0xf2f1c8, 0xf2f1d8, 0xf2f1d8, 0xf2f1e8, 0xf2f1e8, 0xf2f1f8, 0xf2f1f8, 0xf2f208, 0xf2f208, 0xf2f218, 0xf2f218, 0xf2f228, 0xf2f228, 0xf2f238, 
        0xf2f238, 0xf2f248, 0xf2f248, 0xf2f258, 0xf2f258, 0xf2f268, 0xf2f268, 0xf2f278, 0xf2f278, 0xf2f288, 0xf2f288, 0xf2f298, 0xf2f298, 0xf2f2a8, 0xf2f2a8, 0xf2f2b8, 0xf2f2b8, 0xf2f2c8, 0xf2f2c8, 
        0xf2f2d8, 0xf2f2d8, 0xf2f2e8, 0xf2f2e8, 0xf2f2f8, 0xf2f2f8, 0xf2f308, 0xf2f308, 0xf2f318, 0xf2f318, 0xf2f328, 0xf2f328, 0xf2f338, 0xf2f338, 0xf2f348, 0xf2f348, 0xf2f358, 0xf2f358, 0xf2f368, 
        0xf2f368, 0xf2f378, 0xf2f378, 0xf2f388, 0xf2f388, 0xf2f398, 0xf2f398, 0xf2f3a8, 0xf2f3a8, 0xf2f3b8, 0xf2f3b8, 0xf2f3c8, 0xf2f3c8, 0xf2f3d8, 0xf2f3d8, 0xf2f3e8, 0xf2f3e8, 0xf2f3f8, 0xf2f3f8, 
        0xf2f408, 0xf2f408, 0xf2f418, 0xf2f418, 0xf2f428, 0xf2f428, 0xf2f438, 0xf2f438, 0xf2f448, 0xf2f448, 0xf2f458, 0xf2f458, 0xf2f468, 0xf2f468, 0xf2f478, 0xf2f478, 0xf2f488, 0xf2f488, 0xf2f498, 
        0xf2f498, 0xf2f4a8, 0xf2f4a8, 0xf2f4b8, 0xf2f4b8, 0xf2f4c8, 0xf2f4c8, 0xf2f4d8, 0xf2f4d8, 0xf2f4e8, 0xf2f4e8, 0xf2f4f8, 0xf2f4f8, 0xf2f508, 0xf2f508, 0xf2f518, 0xf2f518, 0xf2f528, 0xf2f528, 
        0xf2f538, 0xf2f538, 0xf2f548, 0xf2f548, 0xf2f558, 0xf2f558, 0xf2f568, 0xf2f568, 0xf2f578, 0xf2f578}, large_free_buckets = {0x0 <repeats 17 times>, 0x7ffff7f203c0, 0x0 <repeats 46 times>}, 
      rest_buckets = {0xf2f788, 0xf2f788}, rest_count = 0}
    (gdb) p p
    $17 = (void *) 0x1030f30
    (gdb) p/d p
    $18 = 16977712
    */
    zend_mm_block *mm_block;
    zend_mm_block *next_block;
    size_t size;
#ifdef ZEND_SIGNALS
    TSRMLS_FETCH();
#endif
    /*
    # define ZEND_MM_VALID_PTR(block) \
        zend_mm_check_ptr(heap, block, 1 ZEND_FILE_LINE_RELAY_CC ZEND_FILE_LINE_ORIG_RELAY_CC)
    */
    if (!ZEND_MM_VALID_PTR(p)) {
        return;
    }

    HANDLE_BLOCK_INTERRUPTIONS();

    /*
    #define ZEND_MM_ALIGNMENT               8
    #define ZEND_MM_ALIGNMENT_MASK          ~(ZEND_MM_ALIGNMENT-1)
    #define ZEND_MM_ALIGNED_SIZE(size)      (((size) + ZEND_MM_ALIGNMENT - 1) & ZEND_MM_ALIGNMENT_MASK)
    sizeof(zend_mm_block) = 16
    #define ZEND_MM_ALIGNED_HEADER_SIZE     ZEND_MM_ALIGNED_SIZE(sizeof(zend_mm_block))
    #define ZEND_MM_BLOCK_AT(blk, offset)   ((zend_mm_block *) (((char *) (blk))+(offset)))
    #define ZEND_MM_HEADER_OF(blk)          ZEND_MM_BLOCK_AT(blk, -(int)ZEND_MM_ALIGNED_HEADER_SIZE)
    */
    mm_block = ZEND_MM_HEADER_OF(p);      /* {info = {_size = 16977712, _prev = 49}} */
    /*
    #define ZEND_MM_BLOCK_SIZE(b)           ((b)->info._size & ~ZEND_MM_TYPE_MASK)
    */
    size = ZEND_MM_BLOCK_SIZE(mm_block);  /* 16977712 (0x1030f30) */
    ZEND_MM_CHECK_PROTECTION(mm_block);

#if ZEND_DEBUG || ZEND_MM_HEAP_PROTECTION
    memset(ZEND_MM_DATA_OF(mm_block), 0x5a, mm_block->debug.size);
#endif

#if ZEND_MM_CACHE
    if (EXPECTED(ZEND_MM_SMALL_SIZE(size)) && EXPECTED(heap->cached < ZEND_MM_CACHE_SIZE)) {
        size_t index = ZEND_MM_BUCKET_INDEX(size);
        zend_mm_free_block **cache = &heap->cache[index];

        ((zend_mm_free_block*)mm_block)->prev_free_block = *cache;
        *cache = (zend_mm_free_block*)mm_block;
        heap->cached += size;
        ZEND_MM_SET_MAGIC(mm_block, MEM_BLOCK_CACHED);
#if ZEND_MM_CACHE_STAT
        if (++heap->cache_stat[index].count > heap->cache_stat[index].max_count) {
            heap->cache_stat[index].max_count = heap->cache_stat[index].count;
        }
#endif
        HANDLE_UNBLOCK_INTERRUPTIONS();
        return;
    }
#endif

    /*
    heap->size = 9120
    size = 16977712 (0x1030f30)
    */
    heap->size -= size;
    /*
    heap->size = 18446744073692583024 (0xfffffffffefd1470)
    */

    /*
    #define ZEND_MM_IS_FREE_BLOCK(b)        (!((b)->info._size & ZEND_MM_USED_BLOCK))
    #define ZEND_MM_BLOCK_AT(blk, offset)   ((zend_mm_block *) (((char *) (blk))+(offset)))
    */

    next_block = ZEND_MM_BLOCK_AT(mm_block, size);
    if (ZEND_MM_IS_FREE_BLOCK(next_block)) {  /* SEGV */
        zend_mm_remove_from_free_list(heap, (zend_mm_free_block *) next_block);
        size += ZEND_MM_FREE_BLOCK_SIZE(next_block);
    }
    if (ZEND_MM_PREV_BLOCK_IS_FREE(mm_block)) {
        mm_block = ZEND_MM_PREV_BLOCK(mm_block);
        zend_mm_remove_from_free_list(heap, (zend_mm_free_block *) mm_block);
        size += ZEND_MM_FREE_BLOCK_SIZE(mm_block);
    }
    if (ZEND_MM_IS_FIRST_BLOCK(mm_block) &&
        ZEND_MM_IS_GUARD_BLOCK(ZEND_MM_BLOCK_AT(mm_block, size))) {
        zend_mm_del_segment(heap, (zend_mm_segment *) ((char *)mm_block - ZEND_MM_ALIGNED_SEGMENT_SIZE));
    } else {
        ZEND_MM_BLOCK(mm_block, ZEND_MM_FREE_BLOCK, size);
        zend_mm_add_to_free_list(heap, (zend_mm_free_block *) mm_block);
    }
    HANDLE_UNBLOCK_INTERRUPTIONS();
}
```

I'm not sure exactly how these zend_mm_block things work, but `p` doesn't seem to point to one.  It points to the pthread mutex:

```
(gdb) p gss_mutex
$27 = (void *) 0x1030f30
(gdb) p/d gss_mutex
$28 = 16977712
```

`ZEND_MM_VALID_PTR` doesn't seem to be doing its job.  However, if it did detect that the pointer wasn't to a zend_mm_block, the memory still wouldn't be freed.  My primitive understanding of C tells me that `free` needs to be used instead of `bson_free`.

**Comment by bjori** [comment]
Huge props for you investigation. From skimming your comments, I think phongo is restoring the libbson memory handler to late.

Whatever `mongoc_init()` allocates before phongo calls `bson_mem_set_vtable` needs to be freed using whatever memory manager that was used.

mongoc does however register `atexit()` to cleanup after itself on exit, in case its ´mongoc_cleanup()´ routine isn't called. If that happens though, phongo doesn't restore the memory manager, and will happily attempt to free from the zmm whatever mongoc is trying to free.

So, I'm very curious, how is that exit() triggered? Why is the process straight up exiting rather then hitting the normal shutdown routines?

**Comment by bjori** [comment]
Could you try recompiling mongodb with `CFLAGS=-DMONGOC_NO_AUTOMATIC_GLOBALS`

Something like this should probably work:
```
$ git clone https://github.com/mongodb/mongo-php-driver.git
$ cd mongo-php-driver
$ git submodule update --init
$ phpize
$ CFLAGS=-DMONGOC_NO_AUTOMATIC_GLOBALS ./configure --with-mongodb-sasl
$ make -j4
$ sudo mv /usr/lib/php5/20131226/mongodb.so /usr/lib/php5/20131226/mongodb-original.so
$ sudo cp modules/mongodb.so /usr/lib/php5/20131226/mongodb.so
$ sudo service php5-fpm restart
```

**Comment by Zenexer** [comment]
`MONGOC_NO_AUTOMATIC_GLOBALS` seems to fix it.

`php5-fpm -t` may have an abnormal exit routine because it doesn't actually run any PHP; it just checks that php.ini is valid.  I'm not sure how that's implemented behind-the-scenes, but presumably it could be taking some shortcuts.

There's also the `opcache.fast_shutdown` option.  We've experimented with it a bit and never found it cause problems in the past. so we've left it on.  According to the documentation, that option does interfere with how memory is released (or not released).

**Comment by EugenMayer** [comment]
opcache.fast_shutdown=1 or opcache.fast_shutdown=0 do both not fix / cause this issues, i tested this.

you can use the php daemon mode to test the shutdown 
/usr/bin/php -S localhost:8085

Also you can use valgrind to ensure the leak is fixed:

export USE_ZEND_ALLOC=0
export ZEND_DONT_UNLOAD_MODULES=1
valgrind --leak-check=full /usr/bin/php -S localhost:8085

--
Also this bug is triggered without any PHP script, its the garbage-collector called for every fpm fork.
So you can even reproduce it without stopping fpm, by simply start forks and let them debootstrap again due to max_childs
--

I was yet not able to verify if -DMONGOC_NO_AUTOMATIC_GLOBALS does fix it for us too.

**Comment by Zenexer** [comment]
As @EugenMayer noted, toggling `opcache.fast_shutdown` doesn't seem to affect the problem.  Furthermore, disabling opcache altogether still results in a segfault.

**Comment by EugenMayer** [comment]
@Zenexer for php5 i can confirm that opcache runs pretty stable and is worth any effort ( extraorbitant gain ). Opcache did not have an effect on this issue, neither igbinary - at least not under PHP5.5

For PHP7 things seem to be different, but thats offtopic

**Comment by bjori** [comment]
`opcache.fast_shutdown` isn't related. It does not cause the process to `exit()`, and the memory it cleans up is explicitly allocated by the zend MM so it can't interfere anyway.
AFAICT normal module shutdown still takes place too.

The issue is when a PHP process abnormally `exit()`s without calling the `MSHUTDOWN` sequences of the loaded modules.
This is because mongoc registers `atexit()` handlers to destruct its own loaded libraries, in the even its cleanup routine had not been called (as is the case when you just `exit()` without calling `mongoc_cleanup()`). This however causes a bit of a chicken and egg problem in that the PHP MongoDB extension had overloaded the allocation rountines to use Zend MM, but was never able to relinquish that setting back to normal, as its own `MSHUTDOWN` was never called.


The real fix for this problem is to make sense of why fpm is calling straight up `exit()` rather then the PHP normal `php_module_shutdown()` and `sapi_shutdown()` first for example.
That might be a bug, that might be intentional.
Either way, the fix for the MongoDB extension is to disable the automatic initialization and shutdown routines of mongoc by compiling with  `MONGOC_NO_AUTOMATIC_GLOBALS` as it already does on windows

**Comment by bjori** [comment]
Cross linking with https://jira.mongodb.org/browse/PHPC-632

**Comment by oerdnj** [comment]
@bjori Should I take some measures for creating the mongodb package for `ppa:ondrej/php` users meanwhile? Compiling without libsasl2-dev? Compiling with `CFLAGS=-DMONGOC_NO_AUTOMATIC_GLOBALS` (although I guess this is more a development option, right?)?

**Comment by bjori** [comment]
@oerdnj I would compile mongoc with `CFLAGS=-DMONGOC_NO_AUTOMATIC_GLOBALS` yes. I'd also recommend configuring `libbson` with `--disable-extra-align` if its not to late :)

**Comment by bjori** [comment]
FYI: @oerdnj realizing there was no configure flag for the mongoc magic, I've added a ticket to make it explicit configure switch: https://jira.mongodb.org/browse/CDRIVER-1160

**Comment by sshilko** [comment]
this seems to work for me now, but still testing
{code}
#!/bin/bash

#Compiles and install mongodb extension because pecl produces segfaults and corrupts zend memory heap
#https://github.com/mongodb/mongo-php-driver/issues/209

apt-get install php-pear -y
pecl uninstall mongodb
apt-get remove libsasl2-dev -y
cd /tmp
rm -rf mongo-php-driver
git clone https://github.com/mongodb/mongo-php-driver.git
cd mongo-php-driver
git checkout 1.1.4
git submodule update --init
phpize
make clean
CFLAGS="-DMONGOC_NO_AUTOMATIC_GLOBALS -g -O2 -pipe" ./configure --enable-mongodb
make && make install
/etc/init.d/php7.0-fpm restart
{code}

and
{code}
/usr/sbin/php-fpm7.0 -t
{code}
no longer segfaults

**Comment by EugenMayer** [comment]
maybe also use valgrind to assure its working?

**Comment by sshilko** [comment]
i put it under load, will see metrics later. for me 50 php fpm workers with static pm were ?crashing? every few hrs.. probably when one of them hit max_requests , i tracked it by looking at number of persistent db connects to mysql, dropping by 50 every few hrs..

if nothing drops now, probably its fine.
also no segfaults after '-t' configtests reliably on 3 different machines and different php 7 versions.

happy friday everyone

**Comment by jmikola** [comment]
[1.1.5](https://github.com/mongodb/mongo-php-driver/releases/tag/1.1.5) has been released with the fix for those using bundled libmongoc. Please note that using libmongoc as a system library will require an upstream fix (see: [CDRIVER-1165](https://jira.mongodb.org/browse/CDRIVER-1165)), so folks in that camp may want to switch to using bundled libmongoc for the time being.

**Comment by oerdnj** [comment]
@jmikola Thanks, I am using bundled libmongoc in the package for the moment and eagerly awaiting the version that would drop the -priv usage.

**Comment by jmikola** [comment]
@oerdnj: [PHPC-542](https://jira.mongodb.org/browse/PHPC-542) is the issue to track for that. I'm hoping to have it completed for 1.2.0.

**Comment by Zenexer** [comment]
We haven't encountered any issues since the MONGOC_NO_AUTOMATIC_GLOBALS fix.  We'll switch to the official 1.1.5 release shortly.  Many thanks to everyone who helped solve this quickly.

**Comment by EugenMayer** [comment]
@sshilko @jmikola i just upgraded to 1.1.7 and got the same segmentation faults - right away.
Its seem yet not to be fixed, isnt it?

**Comment by moderndeveloperllc** [comment]
Are you using the bundled or system libmongoc?

**Comment by EugenMayer** [comment]
@moderndeveloperllc using the one you actually use when you use pecl install mongodb under debian wheezy.

This seems to be the exact issue as before. Its only present when you have the headers for sasl and libmemcached presetn. When you remove those and use pecl install mongodb then, 1.1.7 works, but its the same issue as it has been with 1.1.2 and upper.

So the way to avoid it is

```
pecl uninstall memcached
apt-get -y remove libsasl2-dev libmemcached-dev
pecl install mongodb
apt-get -y install libsasl2-dev libmemcached-dev
pecl install memcached
```

But that is obviously no solution at all.

So reproduction ( with docker or what you like ):
debian wheezy (but most probably unrelated)

```
apt-get -y install php5 php5-dev valgrind
apt-get -y install libsasl2-dev libmemcached-dev
pecl install mongodb
valgrind --leak-check=full /usr/bin/php -S localhost:8085
```

**Comment by oerdnj** [comment]
FWIW Debian wheezy has PHP 5.4, so it is a relevant information. Are you able to reproduce it with Debian Jessie?

**Comment by EugenMayer** [comment]
i am using dotdeb

```
deb     http://packages.dotdeb.org wheezy-php55 all
deb-src http://packages.dotdeb.org wheezy-php55 all

php --version
PHP 5.5.33-1~dotdeb+7.1
```

with jessie i did only have php7 here. I am not very eager to test all distro versions / matrix to see were it does not fail. I would say, my setup will be the most practical one you currently will see in a PHP env, maybe php56. So i would tend to say its relevant to fix this combination, no matter if others fail - dont you think?

If i have time, i would spin up some docker image and see what happens.

**Comment by oerdnj** [comment]
Depends on definition of most practical. Wheezy is oldstable (aka LTS)  and PHP 5.5 security support ends in a month. Wheezy also have old(er) cyrus-sasl2 (2.1.25 vs. 2.1.26) and libmemcached (1.0.8 vs 1.0.18) and mongodb might be triggering a bug there.

**Comment by moderndeveloperllc** [comment]
FWIW, @EugenMayer, you really should be using @oerdnj's PPAs :-D By far the most up to date, best thought out repos for Debian/Ubuntu. Just got through installing his Apache and PHP PPAs on a half dozen VMs yesterday. I'll go be quite now..

**Comment by EugenMayer** [comment]
@oerdnj i started with the docker images here: https://github.com/EugenMayer/mongodbtests

valgrind already running on CMD. For the setup there, i could neither reproduce it with jessie or with wheezy (dotdeb php55) yet, but i am still introducing the "stack differences" one by one to see, what is the cause.

**Comment by oerdnj** [comment]
Apart from maintaining PHP in Debian, I also maintain cyrus-sasl2, so I can probably help with pushing fixed version to Debian LTS if the culprit is cyrus-sasl2.

**Comment by EugenMayer** [comment]
@oerdnj thanks!

i move the repo to https://github.com/EugenMayer/php-tests and added a readme, if you want to reproduce.

Well, i could not reproduce the segfault yet, even by imitating the stack one step by one. I did found a different issue https://github.com/mkoppanen/imagick/issues/156 caused by imagick and xdebug - but thats not interesting here.

Only running mongdodb seems not to cause segfaults on this docker images - but on all of my production systems. I am yet not sure what the difference is.

**Comment by EugenMayer** [comment]
I finally could reproduce the issue on the fresh isolated wheezy installation. The important things are 

```
https://github.com/EugenMayer/php-tests/blob/master/mongo/wheezy/Dockerfile#L19
```

so libsasl2-modules-gssapi-mit and libsasl2-dev being present during compiling mongodb with

```
pecl install mongodb
```

## Reproduction using

```
git clone https://github.com/EugenMayer/php-tests
cd php-tests
docker-compose up mongowheezy
docker exec -i -t mongowheezy bash
export USE_ZEND_ALLOC=0
export ZEND_DONT_UNLOAD_MODULES=1
valgrind --leak-check=full /usr/bin/php -S localhost:8085
CMD-c
```

the valgrind output is

```
^C==16130== Warning: set address range perms: large range [0x12dd5000, 0x32dd5000) (noaccess)
==21229== HEAP SUMMARY:
==21229==     in use at exit: 82,968 bytes in 2,608 blocks
==21229==   total heap usage: 23,106 allocs, 20,498 frees, 3,628,170 bytes allocated
==21229==
==21229== 212 (192 direct, 20 indirect) bytes in 1 blocks are definitely lost in loss record 51 of 102
==21229==    at 0x4C28BED: malloc (vg_replace_malloc.c:263)
==21229==    by 0xA4648B2: ???
==21229==    by 0xA4673AE: ???
==21229==    by 0x921D447: sasl_client_add_plugin (in /usr/lib/x86_64-linux-gnu/libsasl2.so.2.0.25)
==21229==    by 0x92291D1: _sasl_load_plugins (in /usr/lib/x86_64-linux-gnu/libsasl2.so.2.0.25)
==21229==    by 0x921DCC0: sasl_client_init (in /usr/lib/x86_64-linux-gnu/libsasl2.so.2.0.25)
==21229==    by 0x8FC42F7: _mongoc_do_init (mongoc-init.c:91)
==21229==    by 0x804E89F: pthread_once (pthread_once.S:104)
==21229==    by 0x8FE17FA: zm_startup_mongodb (php_phongo.c:2471)
==21229==    by 0x6F1515: zend_startup_module_ex (zend_API.c:1759)
==21229==    by 0x6FDA54: zend_hash_apply (zend_hash.c:716)
==21229==    by 0x6F5419: zend_startup_modules (zend_API.c:1886)
```

so it seems sasl related, and again _mongoc_do_init

**Comment by EugenMayer** [comment]
@oerdnj to make this a bit critical - its also reproduceable using the jessie build - so its also broken on the current stable with the php version of current stable : PHP 5.6.20-0+deb8u1
You see the setup here: https://github.com/EugenMayer/php-tests/blob/master/mongo/jessie/Dockerfile

## Reprodcution
```
git clone https://github.com/EugenMayer/php-tests
cd php-tests
docker-compose up mongojessie
docker exec -i -t mongojessie bash
export USE_ZEND_ALLOC=0
export ZEND_DONT_UNLOAD_MODULES=1
valgrind --leak-check=full /usr/bin/php -S localhost:8085
CMD-c
```

```

```
the output will be

```
[Wed Jun  8 23:02:10 2016] Failed to listen on localhost:8085 (reason: Address already in use)
==14==
==14== HEAP SUMMARY:
==14==     in use at exit: 15,122 bytes in 64 blocks
==14==   total heap usage: 26,446 allocs, 26,382 frees, 4,186,644 bytes allocated
==14==
==14== 16 bytes in 1 blocks are definitely lost in loss record 1 of 34
==14==    at 0x4C28C20: malloc (vg_replace_malloc.c:296)
==14==    by 0x7E8EC97: glob (glob.c:452)
==14==    by 0x74955F7: ??? (in /usr/lib/x86_64-linux-gnu/libgssapi_krb5.so.2.2)
==14==    by 0x7495BAA: gss_indicate_mechs (in /usr/lib/x86_64-linux-gnu/libgssapi_krb5.so.2.2)
==14==    by 0x7497935: gss_indicate_mechs_by_attrs (in /usr/lib/x86_64-linux-gnu/libgssapi_krb5.so.2.2)
==14==    by 0xAB8CC95: ???
==14==    by 0xAB8DF8E: ???
==14==    by 0x92B2ED9: sasl_client_add_plugin (in /usr/lib/x86_64-linux-gnu/libsasl2.so.2.0.25)
==14==    by 0x92BEDFF: ??? (in /usr/lib/x86_64-linux-gnu/libsasl2.so.2.0.25)
==14==    by 0x92B37B0: sasl_client_init (in /usr/lib/x86_64-linux-gnu/libsasl2.so.2.0.25)
==14==    by 0x9058077: _mongoc_do_init (mongoc-init.c:91)
==14==    by 0x818944F: pthread_once (pthread_once.S:103)
==14==
==14== 212 (192 direct, 20 indirect) bytes in 1 blocks are definitely lost in loss record 30 of 34
==14==    at 0x4C28C20: malloc (vg_replace_malloc.c:296)
==14==    by 0xAB8CA57: ???
==14==    by 0xAB8DF8E: ???
==14==    by 0x92B2ED9: sasl_client_add_plugin (in /usr/lib/x86_64-linux-gnu/libsasl2.so.2.0.25)
==14==    by 0x92BEDFF: ??? (in /usr/lib/x86_64-linux-gnu/libsasl2.so.2.0.25)
==14==    by 0x92B37B0: sasl_client_init (in /usr/lib/x86_64-linux-gnu/libsasl2.so.2.0.25)
==14==    by 0x9058077: _mongoc_do_init (mongoc-init.c:91)
==14==    by 0x818944F: pthread_once (pthread_once.S:103)
==14==    by 0x90749DA: zm_startup_mongodb (php_phongo.c:2471)
==14==    by 0x6EEC0D: zend_startup_module_ex (in /usr/bin/php5)
==14==    by 0x6FB242: zend_hash_apply (in /usr/bin/php5)
==14==    by 0x6F2CA9: zend_startup_modules (in /usr/bin/php5)
==14==
```

**Comment by EugenMayer** [comment]
an additional hint:

while adding does cause the issue, removing and recompiling does not "fix it", this means, it ist most probably not directly, but rather one of its dependencies? 
![image](https://cloud.githubusercontent.com/assets/136934/15921382/6d0ee7e2-2e21-11e6-8c3b-31646235cf67.png)

**Comment by sshilko** [comment]
@EugenMayer compile with 
CFLAGS=-DMONGOC_NO_AUTOMATIC_GLOBALS

should fix issues (**mongodb.sh**)

```
#!/bin/bash

#Compiles and install mongodb extension because pecl produces segfaults and corrupts zend memory heap
#https://github.com/mongodb/mongo-php-driver/issues/209
#ISSUE IS FIXED IN 1.1.5 (MONGOC_NO_AUTOMATIC_GLOBALS is since 1.1.5 a default compile flag)

apt-get install php-pear -y
pecl uninstall mongodb
apt-get remove libsasl2-dev -y
cd /tmp
rm -rf mongo-php-driver
git clone https://github.com/mongodb/mongo-php-driver.git
cd mongo-php-driver
git checkout 1.1.4
git submodule update --init
phpize
make clean
CFLAGS="-DMONGOC_NO_AUTOMATIC_GLOBALS -g -O2 -pipe" ./configure --enable-mongodb
make && make install
/etc/init.d/php7.0-fpm restart
```

**Comment by EugenMayer** [comment]
@sshilko thanks for this hint!
You do not even need to compile yourself though, as mentioned, before running pecl install, remove libsasl2-dev and you cann compile it with PECL.

 Beside that, i would love to move on with @oerdnj to fix this upstream

**Comment by EugenMayer** [comment]
@oerdnj any way i can help you moving on on this issue?

**Comment by oerdnj** [comment]
@EugenMayer I am responsible only for the `ppa:ondrej/php`, so if you are asking for updated package, I just did that (staging through `ppa:ondrej/php-qa` first). If you are asking this to be fixed in upstream, you should be barking at @jmikola tree...

**Comment by EugenMayer** [comment]
@jmikola bark :) 

@oerdnj just in case, what is this ppa for - is this a alternative for dotdeb or lets say, rather for dotdeps "mongodb" package and an alternative for the PECL one? Means, using your ppa i should be able to have a stable build with jessie/wheezy? If yes, i am happy to try to reproduce that with the docker images i created to reproduce this

**Comment by oerdnj** [comment]
@EugenMayer Forgot about the jessie bit...  I am the Debian Developer behind the PHP, so I have Ubuntu PPAs and recently I started also to publish repository for Debian (https://packages.sury.org/php/). The repositories are not as extensive as dotdeb, but they are prepared in sync of the PHP packaging work I do in Debian unstable.

**Comment by EugenMayer** [comment]
@oerdnj so basically, with jessie i could use https://packages.sury.org/php/ but with wheezy i basically stuck in the middle of nowhere, right?
I see that the jessie php-mongodb builds depend on libc6 (>= 2.14) .. which basically is also broken for the current mongodb-org packages... they are wheezy based, but still depend on libc6 (>= 2.14) .. so 3.2.7 is broken .. and also the meta package is broken.

To be honest, i am currently really rethinking the choice of mongodb in general. The amount of issues is astonishing, is it the php driver, is it the server or its "official state of the art packages". But wel, thats offtopoic.

More or less y/n question: There is no fix for wheezy and there will not be one, because there is not enaugh interest, right?

**Comment by oerdnj** [comment]
@EugenMayer Right, I do not intend to support Debian wheezy. I don't know about dotdeb, but he pulls the packages from unstable, where I just uploaded 1.1.7. Ask there :).

**Comment by jmikola** [comment]
I was traveling for the past few days. I've re-opened this for the moment so it's back on the open issues list and will attempt to dive into it this week.

**Comment by sshilko** [comment]
@EugenMayer i went the same path. got rid of mongodb after having it for 1year in production after my QA mongodb lost the database.

Now sleeping safe with AWS RDS 5.7 mysql json columns.. the migration was easy and painless... just take all data from mongodb and insert into json column in mysql, extracting mongoId as primary varbinary(12) key 🎯 

but ofc. depends on your workloads

mysql 5.7 + innodb_flush_log_at_trx_commit=2 + sync_binlog=0 works for me

**Comment by jmikola** [comment]
@EugenMayer: I started up your docker instance (thanks for putting the effort into creating that, btw) and observed the following memory leak:

```
$ docker exec -i -t mongowheezy bash
root@74e477658bb5:/# export USE_ZEND_ALLOC=0                                                       
root@74e477658bb5:/# export ZEND_DONT_UNLOAD_MODULES=1
root@74e477658bb5:/# valgrind --leak-check=full /usr/bin/php -S localhost:8085
==15== Memcheck, a memory error detector
==15== Copyright (C) 2002-2011, and GNU GPL'd, by Julian Seward et al.
==15== Using Valgrind-3.7.0 and LibVEX; rerun with -h for copyright info
==15== Command: /usr/bin/php -S localhost:8085
==15== 
[Tue Jun 21 14:18:21 2016] Failed to listen on localhost:8085 (reason: Address already in use)
==15== 
==15== HEAP SUMMARY:
==15==     in use at exit: 90,340 bytes in 2,633 blocks
==15==   total heap usage: 26,135 allocs, 23,502 frees, 4,127,091 bytes allocated
==15== 
==15== 212 (192 direct, 20 indirect) bytes in 1 blocks are definitely lost in loss record 49 of 102
==15==    at 0x4C28BED: malloc (vg_replace_malloc.c:263)
==15==    by 0xA61E8B2: ???
==15==    by 0xA6213AE: ???
==15==    by 0x9173447: sasl_client_add_plugin (in /usr/lib/x86_64-linux-gnu/libsasl2.so.2.0.25)
==15==    by 0x917F1D1: ??? (in /usr/lib/x86_64-linux-gnu/libsasl2.so.2.0.25)
==15==    by 0x9173CC0: sasl_client_init (in /usr/lib/x86_64-linux-gnu/libsasl2.so.2.0.25)
==15==    by 0x8F1A2F7: _mongoc_do_init (mongoc-init.c:91)
==15==    by 0x804E89F: pthread_once (pthread_once.S:104)
==15==    by 0x8F377FA: zm_startup_mongodb (php_phongo.c:2471)
==15==    by 0x6F1515: zend_startup_module_ex (zend_API.c:1759)
==15==    by 0x6FDA54: zend_hash_apply (zend_hash.c:716)
==15==    by 0x6F5419: zend_startup_modules (zend_API.c:1886)
==15== 
==15== LEAK SUMMARY:
==15==    definitely lost: 192 bytes in 1 blocks
==15==    indirectly lost: 20 bytes in 2 blocks
==15==      possibly lost: 0 bytes in 0 blocks
==15==    still reachable: 90,128 bytes in 2,630 blocks
==15==         suppressed: 0 bytes in 0 blocks
==15== Reachable blocks (those to which a pointer was found) are not shown.
==15== To see them, rerun with: --leak-check=full --show-reachable=yes
==15== 
==15== For counts of detected and suppressed errors, rerun with: -v
==15== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 41 from 9)
root@74e477658bb5:/#
```

I assume "Failed to listen on localhost:8085" is expected and you wanted this process to terminate quickly, since the docker image (started with `docker-compose up mongowheezy`) ends up running `/usr/bin/php -S localhost:8085` on its own. Is that correct?

I did not see the second leak you reported in https://github.com/mongodb/mongo-php-driver/issues/209#issuecomment-224757443, which includes `libgssapi_krb5.so` in its trace. I don't believe I'm seeing any segfaults via the docker command above (should I be?).

> I see that the jessie php-mongodb builds depend on libc6 (>= 2.14) .. which basically is also broken for the current mongodb-org packages... they are wheezy based, but still depend on libc6 (>= 2.14) .. so 3.2.7 is broken .. and also the meta package is broken.

Note that we do not maintain or publish any distribution packages for the PHP driver. I don't quite follow what you mean by the `mongodb-org` packages being wheezy based. If that involves the packages from MongoDB's own APT repository (see: [here](https://docs.mongodb.com/manual/tutorial/install-mongodb-on-debian/)), it would be something to report in the [SERVER](https://jira.mongodb.org/browse/SERVER) project in MongoDB's JIRA.

**Comment by EugenMayer** [comment]
> @EugenMayer: I started up your docker instance (thanks for putting the effort into creating that, btw)
Fair enough, if your unpaid job is fixing it, my job shall be cracking my head how to reproduce and give you as easy reproduction steps as possible. And i guess, docker is like made for that. So your welcome and more then that, thats how it should be to help you guys crack the harder nuts :)

> I assume "Failed to listen on localhost:8085" is expected and you wanted this process to terminate quickly, since the docker image (started with docker-compose up mongowheezy) ends up running /usr/bin/php -S localhost:8085 on its own. Is that correct?
exactly, by design. Just to quickly bail out and reproduce the memory leak. This valigrind + php -S way was shown to me by a php coredev.

> I did not see the second leak you reported in #209 (comment), which includes libgssapi_krb5.so in its trace. I don't believe I'm seeing any segfaults via the docker command above (should I be?).
Could it be, that you either miss some symbold or maybe need to dsiable zend_alloc?

```
export USE_ZEND_ALLOC=0
export ZEND_DONT_UNLOAD_MODULES=1
```

this should show up then, let me now, if that does not for you

> Note that we do not maintain or publish any distribution packages for the PHP driver. I don't quite follow what you mean by the mongodb-org packages being wheezy based. If that involves the packages from MongoDB's own APT repository (see: here), it would be something to report in the SERVER project in MongoDB's JIRA.

Got that, wont hijack this issue for this issue anymore - sorry.

**Comment by Zenexer** [comment]
Original reporter of the issue here.  There has definitely been a regression.  Interestingly, getting a segfault as pecl exits when installing ext-mongodb 1.1.8 with `sudo -H pecl upgrade mongodb` on Ubuntu 16.04 LTS x86_64.  This appears to interfere with pecl's ability to upgrade other extensions once the error occurs on a system.

Notably, the workaround of disabling SASL support doesn't work this time.

Normally I'd open another issue for a regression, but since there has already been discussion about the matter here, I'll avoid doing that.

```
#0  0x0000000000025a86 in ?? ()
#1  0x00007fffe77f4a88 in zm_shutdown_mongodb (type=<optimized out>, module_number=34) at /tmp/pear/temp/mongodb/php_phongo.c:2527
#2  0x00005555557b80e7 in module_destructor ()
#3  0x00005555557b034c in ?? ()
#4  0x00005555557c3359 in zend_hash_graceful_reverse_destroy ()
#5  0x00005555557b1375 in zend_shutdown ()
#6  0x000055555575214b in php_module_shutdown ()
#7  0x0000555555637cd1 in main ()
```

If I uninstall it and then reinstall it:

```
Program received signal SIGILL, Illegal instruction.
0x00007fffe77f5041 in zm_shutdown_mongodb () from /usr/lib/php/20151012/mongodb.so
(gdb) bt
#0  0x00007fffe77f5041 in zm_shutdown_mongodb () from /usr/lib/php/20151012/mongodb.so
#1  0x00005555557b80e7 in module_destructor ()
#2  0x00005555557b034c in ?? ()
#3  0x00005555557c3359 in zend_hash_graceful_reverse_destroy ()
#4  0x00005555557b1375 in zend_shutdown ()
#5  0x000055555575214b in php_module_shutdown ()
#6  0x0000555555637cd1 in main ()
```

**Comment by Zenexer** [comment]
It's difficult to reproduce the issue reliably without reinstalling the extension, which  makes debugging difficult.  Here's line 2527 in php_phongo.c:

``` c
/* {{{ PHP_MSHUTDOWN_FUNCTION */
PHP_MSHUTDOWN_FUNCTION(mongodb)
{
    (void)type; /* We don't care if we are loaded via dl() or extension= */

    bson_mem_restore_vtable(); /* Line 2527: SEGV */
    /* Cleanup after libmongoc */
    mongoc_cleanup();

    UNREGISTER_INI_ENTRIES();

    return SUCCESS;
}
/* }}} */
```

**Comment by derickr** [comment]
In the past two years, many things have changed. We're closing this ticket, but if the problem still happens, please feel free to open a new ticket.
